pkg/genericapiserver/serialization_config.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package genericapiserver
+
+import (
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type ResourceEncodingConfig interface {
+	// StorageEncoding returns the serialization format for the resource.
+	// TODO this should actually return a GroupVersionKind since you can logically have multiple "matching" Kinds
+	// For now, it returns just the GroupVersion for consistency with old behavior
+	StoragageEncodingFor(unversioned.GroupResource) unversioned.GroupVersion
+
+	// InMemoryEncodingFor returns the groupVersion for the in memory representation the storage should convert to.
+	InMemoryEncodingFor(unversioned.GroupResource) unversioned.GroupVersion
+}
+
+type DefaultResourceEncodingConfig struct {
+	Groups map[string]*GroupResourceEncodingConfig
+}
+
+type GroupResourceEncodingConfig struct {
+	DefaultExternalEncoding   unversioned.GroupVersion
+	ExternalResourceEncodings map[string]unversioned.GroupVersion
+
+	DefaultInternalEncoding   unversioned.GroupVersion
+	InternalResourceEncodings map[string]unversioned.GroupVersion
+}
+
+func NewDefaultResourceEncodingConfig() *DefaultResourceEncodingConfig {
+	return &DefaultResourceEncodingConfig{Groups: map[string]*GroupResourceEncodingConfig{}}
+}
+
+func NewGroupResourceEncodingConfig(defaultEncoding unversioned.GroupVersion) *GroupResourceEncodingConfig {
+	return &GroupResourceEncodingConfig{DefaultExternalEncoding: defaultEncoding, ExternalResourceEncodings: map[string]unversioned.GroupVersion{}}
+}
+
+func (o *DefaultResourceEncodingConfig) SetVersionEncoding(group string, externalEncodingVersion, internalVersion unversioned.GroupVersion) {
+	_, groupExists := o.Groups[group]
+	if !groupExists {
+		o.Groups[group] = NewGroupResourceEncodingConfig(externalEncodingVersion)
+	}
+
+	o.Groups[group].DefaultExternalEncoding = externalEncodingVersion
+	o.Groups[group].DefaultInternalEncoding = internalVersion
+}
+
+func (o *DefaultResourceEncodingConfig) SetResourceEncoding(resourceBeingStored unversioned.GroupResource, externalEncodingVersion, internalVersion unversioned.GroupVersion) {
```

Comment:
It is not clear that the externalVersion is also used as the default external encoding version for the whole group, while internalVersion is not used that way. Why the difference? Is that intentional?